### PR TITLE
Feature/knockback extension

### PR DESCRIPTION
**Pull request by CraZyB1336**
Tweaked some movement settings and hitboxes of the giant fans and super fans.
Air trigger is also now no longer a collider being considered by the projectiles.

Added the knockback extension with a boxing glove model!

The knockback extension works by getting the direction vector from gunController and pushing the enemy its way, and the player the opposite way. Enemy has more force added than the player.

### CODE REVIEW

File: Assets/Scenes/Menu.unity [hunk 1]
```diff
@@ -730,6 +730,22 @@ PrefabInstance:
...
+      value: 
+      objectReference: {fileID: 11400000, guid: 51bb200d04523dd469ed0d8fc736ee64, type: 2}
+    - target: {fileID: 3717336014748129540, guid: 0e2ad36662a02c3488301a5286b27635, type: 3}
+      propertyPath: extensions.Array.size
+      value: 3
+      objectReference: {fileID: 0}
+    - target: {fileID: 3717336014748129540, guid: 0e2ad36662a02c3488301a5286b27635, type: 3}
+      propertyPath: extensions.Array.data[2]
+      value: 
+      objectReference: {fileID: 11400000, guid: 75614ad08cd511f4fa856cd443040455, type: 2}
```
Comment (toberge) [inline Assets/Scenes/Menu.unity:748]: If this part is what I think it is, you should spply the overrides to the StaticInfo prefab. I'd except no changes in the menu scene for this PR.

File: Assets/Scenes/Menu.unity [hunk 2]
```diff
@@ -4093,7 +4109,6 @@ MonoBehaviour:
   defaultMenu: {fileID: 1424215067}
   galleryMenu: {fileID: 1002999806}
   mapNames:
-  - CraterTown
```
Comment (toberge) [inline Assets/Scenes/Menu.unity:4096]: Please revert this :)))))

File: Assets/ScriptableObjects/Items/AuctionStages/ExtensionsWeighted.asset [hunk 3]
```diff
@@ -16,7 +16,7 @@ MonoBehaviour:
   items:
   - {fileID: 11400000, guid: a0f539379f7384445bc1535317457198, type: 2}
   - {fileID: 11400000, guid: bb334224a9209984b98cc009f333ca9d, type: 2}
-  - {fileID: 11400000, guid: bb334224a9209984b98cc009f333ca9d, type: 2}
+  - {fileID: 11400000, guid: 51bb200d04523dd469ed0d8fc736ee64, type: 2}
```
Comment (toberge) [inline Assets/ScriptableObjects/Items/AuctionStages/ExtensionsWeighted.asset:19]: These should be automatically populated by the stuff in StaticInfo, did you change this file directly?

File: ProjectSettings/TagManager.asset [hunk 4]
```diff
@@ -3,7 +3,8 @@
 --- !u!78 &1
 TagManager:
   serializedVersion: 2
-  tags: []
+  tags:
+  - IgnoreCollider
```
Comment (toberge) [inline ProjectSettings/TagManager.asset:7]: Could we use an existing layer for this? Perhaps IgnoreRaycasts?

File: Assets/Scripts/Gamestate/MatchController.cs [hunk 5]
```diff
@@ -101,7 +101,7 @@ void Start()
         // Makes shooting end quickly if testing with 1 player
 #if UNITY_EDITOR
         if (PlayerInputManagerController.Singleton.playerInputs.Count == 1)
-            roundLength = 60f;
+            roundLength = 5f;
```
Comment (toberge) [inline Assets/Scripts/Gamestate/MatchController.cs:104]: Ah, another reminder that we should have a keybind that skips the round in debug mode...

File: Assets/Scripts/Control&Input/PlayerMovement.cs [hunk 6]
```diff
@@ -186,18 +169,15 @@ private void UpdatePosition(Vector3 input)
...
                 }
             case PlayerState.GROUNDED:
                 {
                     // Strafe normally with heavy drag.
                     body.drag = drag;
-                    // Walk along ground normal (adjusted if on heavy slope).
-                    var groundNormal = GroundNormal();
-                    var direction = Vector3.ProjectOnPlane(input, groundNormal);
-                    body.AddForce(direction * strafeForce, ForceMode.VelocityChange);
+                    body.AddForce(input * strafeForce, ForceMode.VelocityChange);
```
Comment (toberge) [inline Assets/Scripts/Control&Input/PlayerMovement.cs:180]: Um, why did you remove the slope movement code? 🤔

File: Assets/Scripts/Control&Input/PlayerMovement.cs [hunk 7]
```diff
@@ -186,18 +169,15 @@ private void UpdatePosition(Vector3 input)
                     // Strafe slightly with less drag.
                     body.drag = airDrag;
                     body.AddForce(input * strafeForceInAir, ForceMode.VelocityChange);
-                    body.AddForce(Vector3.down * MarsGravity, ForceMode.Acceleration);
+                    body.AddForce(Vector3.down * marsGravity * 3.5f, ForceMode.Acceleration);
```
Comment (toberge) [inline Assets/Scripts/Control&Input/PlayerMovement.cs:172]: Why adjust gravity here and not in the constant? And why adjust it in the first place? We changed it to match Mars's gravity (though I do agree that we could turn it up a few notches)

File: Assets/Scripts/Augment/BulletController.cs [hunk 8]
```diff
@@ -34,6 +34,7 @@ protected override void Awake()
     {
         base.Awake();
         UpdateProjectileMovement += ProjectileMotions.MoveWithGravity;
+        bulletsPerShot = bulletsPerShot != 0 ? bulletsPerShot : 1;
```
Comment (toberge) [inline Assets/Scripts/Augment/BulletController.cs:37]: Not sure how this might affect things...

File: Assets/Scripts/Augment/BulletController.cs [hunk 9]
```diff
@@ -14,7 +14,7 @@ public class BulletController : ProjectileController
     private float bulletSpeed = 50f;
 
     [SerializeField]
-    private int bulletsPerShot = 1;
+    public int bulletsPerShot { get; set; }
```
Comment (toberge) [inline Assets/Scripts/Augment/BulletController.cs:17]: Why did you turn it into a property? And please note that properties can't be serialized [according to the Unity docs](https://docs.unity3d.com/ScriptReference/SerializeField.html)

File: Assets/Scripts/Augment/BulletModifiers/KnockbackOnShotModifier.cs [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+    public void Attach(ProjectileController projectile)
+    {
+        projectile.OnProjectileInit += KnockAwayOnShot;
+
+        var bulletController = projectile.gameObject.GetComponent<BulletController>();
+        bulletAmount = bulletController == null ? 1f : bulletController.bulletsPerShot;
+        
+        if (bulletAmount >= 5)
+        {
+            bulletAmount -= 4;
```
Comment (toberge) [inline Assets/Scripts/Augment/BulletModifiers/KnockbackOnShotModifier.cs:36]: You might want to adjust this to adjust more dynamically. We could get weapons with lots more bullets per shot eventually, and have you tested with 2-4 bullets?

File: Assets/Scripts/Augment/BulletModifiers/KnockbackOnShotModifier.cs [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+public class KnockbackOnShotModifier: MonoBehaviour, ProjectileModifier
+{
+    [SerializeField]
+    private float pushPower;
+
+    [SerializeField]
+    private float radius;
+
+    [SerializeField]
+    private KnockbackEffect knockBackScript;
```
Comment (toberge) [inline Assets/Scripts/Augment/BulletModifiers/KnockbackOnShotModifier.cs:17]: Please don't use "script" in names for component references. Everything is a script (or class) here. (don't get me started on the fact that there are C# scripts (csx) now that are a very different thing)

File: Assets/Scripts/Augment/AugmentImplementations/KnockbackProjectileModifier.cs [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+    public void Detach(ProjectileController projectile)
+    {
+        projectile.OnColliderHit -= KnockAwayTarget;
+    }
+
+    public void KnockAwayTarget(Collider collider, ref ProjectileState state)
+    {
+        var stuck = Instantiate(stuckObject, state.position, state.rotation, null);
+        stuck.transform.ParentUnscaled(collider.transform);
+        stuck.gameObject.GetComponent<KnockbackEffect>().KnockAwayTargets(pushPower, knockbackRadius, stuck.transform.position);
```
Comment (toberge) [inline Assets/Scripts/Augment/AugmentImplementations/KnockbackProjectileModifier.cs:34]: So we have two different modifiers for knockback now?

File: Assets/Scripts/Augment/AugmentImplementations/KnockbackEffect.cs [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+        {
+            if (collider.gameObject.TryGetComponent<PlayerManager>(out PlayerManager manager))
+            {
+                var rigidbody = manager.gameObject.GetComponent<Rigidbody>();
+                rigidbody.AddExplosionForce(pushPower, origin, radius, 0f, ForceMode.Impulse);
+            }
+        }
+    }
+
+    public void KnockAwayTargetsDirectional(float pushPower, Vector3 normal, PlayerManager source, float radius)
```
Comment (toberge) [inline Assets/Scripts/Augment/AugmentImplementations/KnockbackEffect.cs:21]: These methods could be static and have a transform passed in. No need to instantiate another object for this... I think?

**Review by toberge — CHANGES_REQUESTED** [review]
Haven't tested the extension in practice yet, but I've found some odd things in the code that you may want to take another look at

**Comment by toberge** [comment]
After testing I notice that

- the shotgun barrel only shoots _one_ bullet with the knockback extension equipped
- the range of the knockback effect seems really low, you have to get _really_ close to trigger it
- the effect might be made a tad stronger?
- the extension model is damn fine :+1:

File: Assets/Scripts/Augment/BulletController.cs (hunk 8 above)
Comment (CraZyB1336) [reply Assets/Scripts/Augment/BulletController.cs:37]: I was thinking if we had a barrel where it outputted no projectiles, the knockback effect would still take place as its own extension.

File: Assets/Scripts/Augment/BulletController.cs (hunk 9 above)
Comment (CraZyB1336) [reply Assets/Scripts/Augment/BulletController.cs:17]: changed it back, but public instead. Found no other way to retrieve bulletspershot info

File: Assets/Scripts/Control&Input/PlayerMovement.cs (hunk 6 above)
Comment (CraZyB1336) [reply Assets/Scripts/Control&Input/PlayerMovement.cs:180]: ???, must've been some merge conflicts, will fix

File: ProjectSettings/TagManager.asset (hunk 4 above)
Comment (CraZyB1336) [reply ProjectSettings/TagManager.asset:7]: Yeah the tag is unused as i used a layer instead, will remove

File: Assets/Scenes/Menu.unity (hunk 1 above)
Comment (CraZyB1336) [reply Assets/Scenes/Menu.unity:748]: Please detail

File: Assets/ScriptableObjects/Items/AuctionStages/ExtensionsWeighted.asset (hunk 3 above)
Comment (CraZyB1336) [reply Assets/ScriptableObjects/Items/AuctionStages/ExtensionsWeighted.asset:19]: No, it changes the population automatically, it probably became like this after testing

File: Assets/Scripts/Augment/BulletController.cs (hunk 8 above)
Comment (toberge) [reply Assets/Scripts/Augment/BulletController.cs:36]: Shouldn't this logic be in the extension code then?

File: Assets/Scripts/Augment/BulletController.cs (hunk 9 above)
Comment (toberge) [reply Assets/Scripts/Augment/BulletController.cs]: You could do a property with a private backing field (i.e., both of these kinda)

File: Assets/Scenes/Menu.unity (hunk 1 above)
Comment (toberge) [reply Assets/Scenes/Menu.unity:748]: image
If you add an extension in **1** you need to apply the overrides by following **2** and **3** (yeah the 2 is badly drawn sorry)

File: Assets/Scripts/Augment/BulletController.cs (hunk 9 above)
Comment (toberge) [reply Assets/Scripts/Augment/BulletController.cs]: Though I'm not sure if this is is something that is actually serialized?

File: Assets/Scenes/Menu.unity (hunk 1 above)
Comment (toberge) [reply Assets/Scenes/Menu.unity:748]: image
I did check if this was the case rn and seems like it

File: Assets/Prefabs/GunParts/RocketLauncherBarrel.prefab
Comment (toberge) [inline Assets/Prefabs/GunParts/RocketLauncherBarrel.prefab:1]: This doesn't seem to be in use, wanna remove it?

File: Assets/Prefabs/GunParts/Bullets/RocketSphere.prefab
Comment (toberge) [inline Assets/Prefabs/GunParts/Bullets/RocketSphere.prefab:1]: Same goes for this one

File: Assets/Scenes/Menu.unity (hunk 1 above)
Comment (CraZyB1336) [reply Assets/Scenes/Menu.unity:748]: i can double check it, sorry for slow pull request review on my end

File: Assets/Scripts/Augment/BulletController.cs (hunk 8 above)
Comment (CraZyB1336) [reply Assets/Scripts/Augment/BulletController.cs:36]: Yes, on further inspection

File: Assets/Scripts/Augment/BulletModifiers/KnockbackOnShotModifier.cs (hunk 10 above)
Comment (CraZyB1336) [reply Assets/Scripts/Augment/BulletModifiers/KnockbackOnShotModifier.cs:36]: Replaced with `(pushPower / bulletAmount) * (1f + (float) Math.Log10(bulletAmount))`. This seems to solve the issue at least for shotgun / bulletamounts, although I'm unsure how we should resolve the "propeller" effect when combining solar body and normal gun barrel.

File: Assets/Prefabs/GunParts/RocketLauncherBarrel.prefab
Comment (CraZyB1336) [reply Assets/Prefabs/GunParts/RocketLauncherBarrel.prefab:1]: Sure, ill keep it somewhere locally, but for the repo we can remove it.

File: Assets/Scripts/Augment/AugmentImplementations/KnockbackProjectileModifier.cs (hunk 12 above)
Comment (CraZyB1336) [reply Assets/Scripts/Augment/AugmentImplementations/KnockbackProjectileModifier.cs:34]: First implementation was for a spherical explosive knockback. The second is just two directional. In the case someone wants to use the first one, for instance a rocket launcher, they could use this method.

File: Assets/Scripts/Augment/AugmentImplementations/KnockbackEffect.cs (hunk 13 above)
Comment (CraZyB1336) [reply Assets/Scripts/Augment/AugmentImplementations/KnockbackEffect.cs:21]: Removed the file, and used gunController's forward vector instead.

File: Assets/Scenes/Menu.unity (hunk 1 above)
Comment (CraZyB1336) [reply Assets/Scenes/Menu.unity]: Its good.

File: Assets/Scripts/Augment/BulletController.cs (hunk 9 above)
Comment (CraZyB1336) [reply Assets/Scripts/Augment/BulletController.cs]: Fredrik showed me the `public fieldName => readFromField` method of reading private fields. Used this to retrieve the information.

File: Assets/Scripts/Augment/AugmentImplementations/KnockbackProjectileModifier.cs (hunk 12 above)
Comment (CraZyB1336) [reply Assets/Scripts/Augment/AugmentImplementations/KnockbackProjectileModifier.cs]: Removed the file and stored it locally. It was unnecessary after some discussion with Fredrik.

File: Assets/Prefabs/GunParts/Bullets/RocketSphere.prefab
Comment (CraZyB1336) [reply Assets/Prefabs/GunParts/Bullets/RocketSphere.prefab:1]: 😎👍

File: Assets/Scripts/Augment/BulletModifiers/KnockbackOnShotModifier.cs (hunk 10 above)
Comment (toberge) [reply Assets/Scripts/Augment/BulletModifiers/KnockbackOnShotModifier.cs]: We'll have to adjust the solar body anyways, but perhaps you could assist knockback effect based on rate of fire?

File: Assets/Scripts/Augment/AugmentImplementations/KnockbackProjectileModifier.cs (hunk 12 above)
Comment (toberge) [reply Assets/Scripts/Augment/AugmentImplementations/KnockbackProjectileModifier.cs]: Mkay

File: Assets/Scripts/Control&Input/PlayerMovement.cs (hunk 7 above)
Comment (CraZyB1336) [reply Assets/Scripts/Control&Input/PlayerMovement.cs]: removed it. It's back to how it was

**Review by toberge — APPROVED** [review]
Works quite well now! The push force might need to be fine-tuned later, but it seems to allow for ~~rocket~~ fist jumping while not propelling you indefinitely upwards with the solar body.

The single merge conflict should be resolved by accepting both changes since another extension has been added.

File: Assets/Scripts/Augment/BulletModifiers/KnockbackOnShotModifier.cs [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+    void Awake()
+    {
+        gunController = transform.parent.GetComponent<GunController>();
+        if (!gunController)
+        {
+            Debug.Log("Fire not attached to gun parent!");
+            return;
+        }
+    }
+public void Attach(ProjectileController projectile)
```
[the remark this answers is not in the document]
Comment (CraZyB1336) [reply Assets/Scripts/Augment/BulletModifiers/KnockbackOnShotModifier.cs]: woopsie :3

[the remark this answers is not in the document]
Comment (CraZyB1336) [reply Assets/Scripts/Augment/BulletModifiers/KnockbackOnShotModifier.cs]: Ill just resolve this since it was minuscule.
